examples/basic_bitcoin/src/backend/bitcoin_wallet.ts

```diff
@@ -0,0 +1,295 @@
+    // and sign a transaction, see what its size is, and then update the fee,
+    // rebuild the transaction, until the fee is set to the correct amount.
+    let totalFee = 0n;
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
+        let transaction = buildTransactionWithFee(
+            ownUtxos,
+            ownAddress,
+            dstAddress,
+            amount,
+            totalFee,
+            network,
+            transactionHashes
+        );
+
+        // Sign the transaction. In this case, we only care about the size
+        // of the signed transaction.
+        const signedTransaction = await signTransaction(
+            ownPublicKey,
+            transaction.clone(),
+            keyName,
+            derivationPath
+        );
+
+        const signedTxBytesLen = BigInt(signedTransaction.byteLength());
+
+        if ((signedTxBytesLen * feePerByte) / 1_000n === totalFee) {
+            return transaction;
+        } else {
+            totalFee = (signedTxBytesLen * feePerByte) / 1_000n;
+        }
+    }
+}
+
+function buildTransactionWithFee(
+    ownUtxos: Utxo[],
+    ownAddress: string, // TDO what about that ownAddress?? Would we only need it for the signing?
+    destAddress: string,
+    amount: bigint,
+    fee: bigint,
+    network: bitcoin.Network,
+    transactionHashes: TransactionHashes
+): Psbt {
+    // Assume that any amount below this threshold is dust.
+    const dustThreshold = 1_000n;
+
+    // Select which UTXOs to spend. We naively spend the oldest available UTXOs,
+    // even if they were previously spent in a transaction. This isn't a
+    // problem as long as at most one transaction is created per block and
+    // we're using min_confirmations of 1.
+    let utxosToSpend: Utxo[] = [];
+    let totalSpent = 0n;
+    for (const utxo of [...ownUtxos].reverse()) {
+        if (utxosToSpend.includes(utxo)) {
+            continue;
+        }
+        totalSpent += utxo.value;
+        utxosToSpend.push(utxo);
+        if (totalSpent >= amount + fee) {
+            // We have enough inputs to cover the amount we want to spend.
+            break;
+        }
+    }
+
+    if (totalSpent < amount + fee) {
+        throw new Error(
+            `Insufficient balance: ${totalSpent}, trying to transfer ${amount} satoshi with fee ${fee}`
+        );
+    }
+
+    let transaction = new Psbt({ network });
+    transaction.setVersion(2);
+
+    for (let utxo of utxosToSpend) {
+        const hash = Buffer.from(utxo.outpoint.txid).reverse();
+        const txid = hash.toString('hex');
+        const nonWitnessUtxo = Buffer.from(transactionHashes[txid], 'hex');
+        transaction.addInput({
+            hash: Buffer.from(utxo.outpoint.txid), // TOOD if this works clean it up
+            index: utxo.outpoint.vout,
+            nonWitnessUtxo
+        });
+    }
+
+    const remainingAmount = totalSpent - (amount + fee);
```

Comment:
I think it's mathematically the same, but can we not do all minuses like in Rust?